src/types.rs

```diff
@@ -450,6 +459,52 @@
     }
 }
 
+pub struct SubjectBuilder {
+    tokens: Vec<String>
+}
+
+impl SubjectBuilder {
+    pub fn new() -> Self {
+        SubjectBuilder {
+            tokens: Vec::new()
+        }
+    }
+
+    pub fn add(mut self, subject: String) -> Self {
+        // Need to add some checks here to check for illegal characters
+        self.tokens.push(subject);
+        self
+    }
+
+    pub fn add_wildcard(mut self) -> Self {
+        self.tokens.push("*".to_string());
+        self
+    }
+
+    pub fn add_full_wildcard(self) -> FullWildcardSubjectBuilder {
```

Comment:
What about making two methods `full_wildcard` that just makes the `Subject` a full wildcard subject and `set_full_wildcard` that takes a boolean and sets the wildcard behavior accordingly.

I think we could then do away with the `FullWildcardSubjectBuilder`.

Reply:
Hmmm, I actually wrote it this way so that making a subject a full wildcard subject is terminal and no tokens can be added after, and it also forces the user to build the subject as he would normally - with the full wildcard as the terminal token, but I suppose this is fine too. I'll make the changes.

Reply:
That makes sense. Could we have a `build_full_wildcard` that would be terminal, but  would avoid adding a new type?

Reply:
Sure, made the changes.